Gelu bw fused operation

philei-tt opened the pull request:
### Ticket
[Github Issue](https://github.com/tenstorrent/tt-metal/issues/14160)

### Problem description
Composite backward gelu operation is slow

### What's changed
Added experimental fused gelu_bw operation. All steps are done in one compute kernel. 

#### Performance
```
[appr=tanh]: Average gelu_bw time over 10000 runs: 0.002072 seconds
[appr=none]: Average gelu_bw time over 10000 runs: 0.001255 seconds
[appr=tanh]: Average experimental gelu_bw time over 10000 runs: 0.000262 seconds
[appr=none]: Average experimental gelu_bw time over 10000 runs: 0.000398 seconds
```

- x7.91 speed up for tanh approximation
- x3.15 speed up for none approximation

#### nano gpt performance:
Composite gelu bw (approximation=none): (~265ms)
```
Step: 100, Loss: 2.609375
Full step time 264.138 ms, cache entries: 96
Step: 200, Loss: 2.484375
Full step time 264.74 ms, cache entries: 96
Step: 300, Loss: 2.390625
Full step time 265.854 ms, cache entries: 96
Step: 400, Loss: 2.265625
Full step time 264.017 ms, cache entries: 96
```

Fused gelu bw (approximation=none): (~238ms)

```
Step: 100, Loss: 2.59375
Full step time 238.756 ms, cache entries: 91
Step: 200, Loss: 2.484375
Full step time 238.687 ms, cache entries: 91
Step: 300, Loss: 2.375
Full step time 238.825 ms, cache entries: 91
Step: 400, Loss: 2.25
Full step time 239.163 ms, cache entries: 91
```

Fused gelu bw (approximation=tanh): (~232ms)

```
Step: 100, Loss: 2.59375
Full step time 232.129 ms, cache entries: 91
Step: 200, Loss: 2.46875
Full step time 231.689 ms, cache entries: 91
Step: 300, Loss: 2.375
Full step time 232.283 ms, cache entries: 91
Step: 400, Loss: 2.21875
Full step time 232.272 ms, cache entries: 91
```


#### Accuracy
```
[GELU_BW pcc] [appr: tanh] experimental: (True, '0.9991919767258448'), composite: (True, '0.9991965176463199')
[GELU_BW pcc] [appr: none] experimental: (True, '0.9999788271028575'), composite: (True, '0.9999835901831405')
```

### Notes
- Performance can probably be improved a lot by adding a proper gelu_bw llk function instead of calling many operations in compute kernel

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

Commit:
[gelu_bw] Initial compilable boilerplate for new op

Commit:
Small native gelu_bw fixes:

- add to api.rst
- fix prim operation path
- Add golden function

Commit:
Fix boilerplate code: make it "binary"

Commit:
Add working bw_gelu compute kernel

Commit:
Add none approximation

Commit:
Unify experiment gelu_bw interface with composite

Commit:
Add documentation of experimental gelu_bw

Commit:
Add unittest for fused gelu bw

Commit:
Add queue_id argument to fused gelu_bw

Commit:
Remove code dublication in fused gelu_bw compute kernel

Commit:
Cleanup (fused gelu bw)

ayerofieiev-tt commented on ttnn/CMakeLists.txt:
```diff
@@ -727,6 +727,9 @@ set(TTNN_OP_SRCS
     ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/experimental/dropout/device/dropout_program_factory.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/experimental/dropout/dropout.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/experimental/reshape/view.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/experimental/unary_backward/unary_backward.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/gelu_backward_device_operation.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/gelu_backward_program_factory.cpp
```
why is folder `unary_backward`?

philei-tt replied:
Since all backward unary operations are composite right now, I thought we would like to make fused variant for all of them eventually, especially since now we have a good example of how to do that, it can be done quite quickly.

philei-tt replied:
Moved in a gelu_backward subdirectory

Denys88 commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/gelu_backward_program_factory.hpp:
```diff
@@ -0,0 +1,30 @@
+        tt::tt_metal::KernelHandle gelu_bw_reader_kernel_id;
+        tt::tt_metal::KernelHandle gelu_bw_compute_kernel_id;
+        tt::tt_metal::KernelHandle gelu_bw_writer_kernel_id;
+        uint32_t num_cores;
```
If initialize it with zeros.

philei-tt replied:
Sorry, didn't get that, do you suggest to initialize `num_cores` and `num_cores_y` to 0 by default?

ayerofieiev-tt replied:
uint32_t num_cores = 0

ayerofieiev-tt replied:
https://github.com/tenstorrent/tt-metal/blob/main/contributing/BestPractices.md#15-initialize-primitive-types-on-declaration

philei-tt replied:
Done

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/unary_backward.hpp:
```diff
@@ -0,0 +1,32 @@
+        QueueId queue_id,
+        const Tensor& grad_tensor_arg,
+        const Tensor& input_tensor_arg,
+        const string& parameter_a,
```
please give argument a good name

ayerofieiev-tt replied:
like `approximation_mode`

Denys88 commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/kernels/compute/eltwise_bw_gelu_approx_none.cpp:
```diff
@@ -0,0 +1,109 @@
+    uint32_t per_core_block_cnt = get_arg_val<uint32_t>(0);
+    uint32_t per_core_block_size = get_arg_val<uint32_t>(1);
+
+    constexpr auto cb_in0 = tt::CBIndex::c_0;   // grad
```
lets call it cb_grad.
And update all other var names.

dgomezTT commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/gelu_backward_device_operation.cpp:
```diff
@@ -0,0 +1,161 @@
+
+void GeluBackwardDeviceOperation::validate_on_program_cache_miss(
+    const operation_attributes_t& args, const tensor_args_t& tensor_args) {
+    // TODO: implement this
```
a todo should always be a ticket!

philei-tt replied:
Oops missed that :)

Denys88 commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/kernels/compute/eltwise_bw_gelu_approx_none.cpp:
```diff
@@ -0,0 +1,109 @@
+
+    constexpr uint32_t bits_1p0 = 0x3f800000;          // 1.0f
+    constexpr uint32_t bits_0p5 = 0x3f000000;          // 0.5f
+    constexpr uint32_t bits_inv_sqrt2 = 0x3f3504f3;    // ≈ 0.70710677 (1 / sqrt(2))
```
How did you calculate this numbers? Is ther  any chance to write something like  constexpr uint32_t bits_1p0 = bitcast(1.0f); ?

patrickroberts replied:
With [`<numbers>`](https://en.cppreference.com/w/cpp/numeric/constants), maybe `constexpr auto bits_inv_sqrt2 = std::bit_cast<uint32_t>(1 / std::numbers::sqrt2_v<float>);`

patrickroberts replied:
https://godbolt.org/z/PMnhqdafP

patrickroberts replied:
https://godbolt.org/z/rf3aa4M7e

philei-tt replied:
@Denys88 @patrickroberts  Hm, it seems like kernel code uses c++17, but bit_cast was introduced in c++20. Is there any way of changing that, or should I dismiss this suggestion?

patrickroberts replied:
Oh, I thought that this was on host side, didn't realize it was kernel code. You should dismiss the suggestion.

philei-tt replied:
Apparently It's not possible right now, so I will resolve this conversation

dmakoviichuk-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/kernels/compute/eltwise_bw_gelu_approx_none.cpp:
```diff
@@ -0,0 +1,109 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
```
2025

dmakoviichuk-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/kernels/compute/eltwise_bw_gelu_approx_none.cpp:
```diff
@@ -0,0 +1,109 @@
+    constexpr uint32_t bits_neg_0p5 = 0xbf000000;      // -0.5f
+
+    unary_op_init_common(cb_in0, cb_out0);
+    erf_tile_init();
```
in binary or unary op chaining they call init inside of the tile cycle. Are you sure tile init calls dont break each other.

philei-tt replied:
I'm not very familiar with what exactly init functions do, but results are correct, no hungs (so far) and no dataraces.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/gelu_backward_device_operation.cpp:
```diff
@@ -0,0 +1,161 @@
+
+tensor_return_value_t GeluBackwardDeviceOperation::create_output_tensors(
+    const operation_attributes_t& args, const tensor_args_t& tensor_args) {
+    if (tensor_args.preallocated_input_grad.has_value()) {
```
This method should return a preallocated output tensor if user provided one.
is `preallocated_input_grad` a good name?

dchenTT replied:
it seems name of input/output reflects there are backwards operation?

philei-tt replied:
Yes, arguments names are similar to the [composite gelu bw](https://github.com/tenstorrent/tt-metal/blob/61b6d27d82cb74c9143e81914cff11f90899b943/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.hpp#L577C1-L578C1)

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/gelu_backward_device_operation.cpp:
```diff
@@ -0,0 +1,161 @@
+        static_cast<int>(input_tensor.get_dtype()),
+        static_cast<int>(output_datatype));
+
+    auto arch = input_tensor.device()->arch();
+    TT_FATAL(
+        arch == tt::ARCH::WORMHOLE_B0,
+        "GELU_BW operation is only supported on Wormhole. Device arch: {}",
+        magic_enum::enum_name(arch));
```
maybe can be removed. mostly existed because of grayskull, but its now deprecated. 
unless you tested and it does not work on BH.

patrickroberts replied:
Also should just be able to pass `arch` instead of `magic_enum::enum_name(arch)`

philei-tt replied:
Okay, I will remove that assert

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/gelu_backward_device_operation.cpp:
```diff
@@ -0,0 +1,161 @@
+        static_cast<int>(input_tensor.memory_config().memory_layout),
+        static_cast<int>(out_memory_config.memory_layout));
+
+    if (!input_tensor.is_sharded()) {
+        TT_FATAL(
+            input_tensor.get_layout() == Layout::TILE,
+            "GELU_BW operation requires tensor to be in Tile layout when working with non-sharded input tensor. Input "
+            "tensor layout: {}",
+            static_cast<int>(input_tensor.get_layout()));
+
+        TT_FATAL(
+            input_tensor.memory_config().memory_layout == TensorMemoryLayout::INTERLEAVED,
+            "GELU_BW operation requires Interleaved memory layout when working with non-sharded input tensor. Input "
+            "memory layout: `{}`",
+            static_cast<int>(input_tensor.memory_config().memory_layout));
+    }
```
```suggestion
      TT_FATAL(
          input_tensor.get_layout() == Layout::TILE,
          "GELU_BW operation requires tensor to be in Tile layout when working with non-sharded input tensor. Input "
          "tensor layout: {}",
          static_cast<int>(input_tensor.get_layout()));

      TT_FATAL(
          input_tensor.memory_config().memory_layout == TensorMemoryLayout::INTERLEAVED,
          "GELU_BW operation requires Interleaved memory layout when working with non-sharded input tensor. Input "
          "memory layout: `{}`",
          static_cast<int>(input_tensor.memory_config().memory_layout));
```

rfurko-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/kernels/compute/eltwise_bw_gelu_approx_none.cpp:
```diff
@@ -0,0 +1,109 @@
+        cb_wait_front(cb_in1, per_core_block_size);
+        cb_reserve_back(cb_out0, per_core_block_size);
+
+        tile_regs_acquire();
```
Please take a look at this issue
https://github.com/tenstorrent/tt-metal/issues/5868#issuecomment-2101726935

basically this two function calls are identical to deprecated call of `acquire_dst` (same for `release_dst`) and it has some issue that it's hard to reproduce. Please double check this.

ayerofieiev-tt replied:
@rdjogoTT @rtawfik01 , wonder wdyt

rdjogoTT replied:
Yea the 4 `tile_regs_*` calls need to be moved into the for loop because they serve to hand off the Dest reg back and forth between MATH and PACK.

philei-tt replied:
It doesn't really matter in this case, since `per_core_block_size` is hardcoded to 1 in the kernel setup, but yeah, that might cause problems in the future, so I will move them in the loop.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/gelu_backward_device_operation_types.hpp:
```diff
@@ -0,0 +1,24 @@
+    std::optional<Tensor> preallocated_input_grad;
+};
+
+using tensor_return_value_t = Tensor;
```
shouldn't it be a `vector<Tensor>`?

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/unary_backward.cpp:
```diff
@@ -0,0 +1,39 @@
+    auto output_memory_config = input_grad_tensor.has_value() ? input_grad_tensor.value().memory_config()
+                                                              : memory_config.value_or(input_tensor.memory_config());
+
+    return {std::optional<Tensor>(ttnn::prim::gelu_bw(
+        queue_id,
+        grad_output_tensor,
+        input_tensor,
+        approximate,
+        output_dtype,
+        output_memory_config,
+        input_grad_tensor))};
```
```suggestion
    return std::optional<Tensor>(ttnn::prim::gelu_bw(
        queue_id,
        grad_output_tensor,
        input_tensor,
        approximate,
        output_dtype,
        output_memory_config,
        input_grad_tensor));
```

ayerofieiev-tt replied:
It will after you will switch from vector<Tensor> to Tensor return type as discussed offline

philei-tt replied:
Done

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/unary_backward.cpp:
```diff
@@ -0,0 +1,39 @@
+        input_grad_tensor))};
+}
+
+OptionalTensors GeluBackwardOperation::create_async_optional_output_tensors(
+    QueueId queue_id,
+    const Tensor& grad_output_tensor,
+    const Tensor& input_tensor,
+    const string& approximate,
+    const std::optional<MemoryConfig>& memory_config,
+    std::optional<Tensor> input_grad_tensor) {
+    return {
+        std::optional<Tensor>(tt::tt_metal::operation::get_workers_for_op_output({grad_output_tensor, input_tensor}))};
+}
```
only needed if op returns a vector<Tensor>

philei-tt replied:
I know, but it has to return a list of tensors to have exactly the same interface as [composite gelu_bw variant](https://github.com/tenstorrent/tt-metal/blob/2b0220986fea23cf05c1d8f1f0de7843a658b243/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.hpp#L577)

ayerofieiev-tt replied:
discussed offline, should be removed

philei-tt replied:
Done

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/unary_backward.cpp:
```diff
@@ -0,0 +1,39 @@
+
+namespace ttnn::operations::experimental {
+
+OptionalTensors GeluBackwardOperation::invoke(
```
```suggestion
Tensor GeluBackwardOperation::invoke(
```

philei-tt replied:
It has to return a vector of tensors to have exactly the same interface as [composite gelu_bw variant](https://github.com/tenstorrent/tt-metal/blob/2b0220986fea23cf05c1d8f1f0de7843a658b243/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.hpp#L577)

ayerofieiev-tt replied:
discussed offline, should be changed here

philei-tt replied:
Done

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/unary_backward.hpp:
```diff
@@ -0,0 +1,32 @@
+namespace ttnn::operations::experimental {
+
+struct GeluBackwardOperation {
+    static OptionalTensors invoke(
```
```suggestion
    static Tensor invoke(
```

philei-tt replied:
It has to return a vector of tensors to have exactly the same interface as [composite gelu_bw variant](https://github.com/tenstorrent/tt-metal/blob/2b0220986fea23cf05c1d8f1f0de7843a658b243/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.hpp#L577)

ayerofieiev-tt replied:
composite unary backwards operations are crap. thats why you are rewriting.

philei-tt replied:
Done

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/unary_backward.hpp:
```diff
@@ -0,0 +1,32 @@
+        const std::optional<MemoryConfig>& memory_config = std::nullopt,
+        std::optional<Tensor> input_grad = std::nullopt);
+
+    static OptionalTensors create_async_optional_output_tensors(
+        QueueId queue_id,
+        const Tensor& grad_output_tensor,
+        const Tensor& input_tensor,
+        const string& approximate,
+        const std::optional<MemoryConfig>& memory_config = std::nullopt,
+        std::optional<Tensor> input_grad_tensor = std::nullopt);
+};
```
can remove. only needed if invoke must return a vector

philei-tt replied:
It has to return a vector of tensors to have exactly the same interface as [composite gelu_bw variant](https://github.com/tenstorrent/tt-metal/blob/2b0220986fea23cf05c1d8f1f0de7843a658b243/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.hpp#L577)

philei-tt replied:
Removed

dmakoviichuk-tt reviewed (approved):
Overall looks really good!
there is the only one place where we use gelu_bw is our tt-train nanogpt training.
As I remember  default performance is near 275ms step.
If you can run it and compare vs your 2 both cases. One run takes ~15-20min.
you can demonstrate perf improvements and make sure loss is the same.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/unary_backward.cpp:
please rename to gelu_backward.cpp

ayerofieiev-tt replied:
and please rename the folder too

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/gelu_backward_device_operation.cpp:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    auto output_layout = Layout::TILE;
+    if (args.output_memory_config.is_sharded()) {
```
but do you truly support sharded?

philei-tt replied:
Nope 😢  Should I add sharded support?

ayerofieiev-tt replied:
No, not now
Just assert that input is not sharded

philei-tt replied:
Added assert

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/gelu_backward_device_operation.cpp:
```diff
@@ -0,0 +1,161 @@
+    const auto& input_shape = input_tensor.get_padded_shape();
+
+    auto program_factory = select_program_factory(args, tensor_args);
+    operation::Hash hash = operation::hash_operation<GeluBackwardDeviceOperation>(
+        args,
+        program_factory.index(),
+        input_tensor.dtype(),
+        std::get<DeviceStorage>(input_tensor.storage()).memory_config(),
+        grad_output.dtype(),
+        std::get<DeviceStorage>(grad_output.storage()).memory_config(),
+        input_shape.volume());
```
these all affect compile-time args of forming of the kernel itself?

philei-tt replied:
Yes, I think so

dchenTT commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/gelu_backward_program_factory.cpp:
```diff
@@ -0,0 +1,202 @@
+
+GeluBackwardProgramFactory::cached_program_t GeluBackwardProgramFactory::create(
+    const operation_attributes_t& args, const tensor_args_t& tensor_args, tensor_return_value_t& output) {
+    using namespace tt;
+    using namespace tt::tt_metal;
```
these two lines are redundant now

dchenTT commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/kernels/compute/eltwise_bw_gelu_approx_none.cpp:
```diff
@@ -0,0 +1,109 @@
+namespace NAMESPACE {
+
+void MAIN {
+    uint32_t per_core_block_cnt = get_arg_val<uint32_t>(0);
+    uint32_t per_core_block_size = get_arg_val<uint32_t>(1);
```
Is there a reason to call it block, when it is actually a tile (both from program factory and actual compute? Is block more general purpose?

ayerofieiev-tt replied:
I think it is generally assumed that one core might be processing multiple blocks. Block might have more than 1 tile. 
Isn't that right?

philei-tt replied:
One core might processes multiple blocks consisting of multiple tiles

dchenTT commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/unary_backward.hpp:
```diff
@@ -0,0 +1,32 @@
+        const std::optional<MemoryConfig>& memory_config = std::nullopt,
+        std::optional<Tensor> input_grad = std::nullopt);
+
+    static OptionalTensors create_async_optional_output_tensors(
```
why it is called async?

philei-tt replied:
Not sure, this is a name of function [required](https://github.com/tenstorrent/tt-metal/blob/3f1a62a7977b095bb638d24c182434f20cde906a/ttnn/cpp/ttnn/decorators.hpp#L63) by operation invoke wrapper when registering it, @ayerofieiev-tt can you explain?

ayerofieiev-tt replied:
it is to comply with an infra.
i hope we will remove this from infra.

also this will be removed from this op in this PR.

ayerofieiev-tt replied:
this is not needed if an op is registered with `register_operation`

But when it is registered with `register_operation_with_auto_launch_op` infra that makes the function call happen in a different thread must know how many tensors to immediately return.

philei-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/device/gelu_backward/kernels/compute/eltwise_bw_gelu_approx_none.cpp:
```diff
@@ -0,0 +1,109 @@
+        tile_regs_wait();
+
+        for (uint32_t i = 0; i < per_core_block_size; ++i) {
+            copy_tile(cb_in0, i, 0);  // grad => tile 0
```
I don't see a good way of doing this. Most tiles keep temporary values during computation, so not sure can we name them. I know it's ugly, so I did my best with comments

Commit:
PR Cleanup: Rename vars. Add copyrights, Remove TODOs

Commit:
Make gelu_bw return Tensor instead of OptionalTensors

Commit:
Remove one obsolete multiplication in gelu_bw tanh compute kernel

Commit:
Change gelu_bw folders structure

philei-tt commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/unary_backward.hpp:
Done

Commit:
A few small changes to gelu_bw:

1. Assert input is not sharded
2. Change loading immediate value type from fp16a to fp16b
3. initialize num_cores with 0

philei-tt commented:
Apparently the accuracy difference is big enough to mess up training of nano gpt: 
composite: Step: 400, Loss: 2.265625
experimental: Step: 400, Loss: 3.296875

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/gelu_backward/gelu_backward_pybind.cpp:
```diff
@@ -0,0 +1,83 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
2025 Tenstorrent AI ULC

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/experimental/unary_backward/gelu_backward/device/kernels/compute/eltwise_bw_gelu_common.hpp:
```diff
@@ -0,0 +1,72 @@
+    }
+}
+
+inline void copy_value(uint dst_index0, uint32_t dst_index1) {
```
Please add a detailed description of this API, what it does and what are the parameters in a comment block.

philei-tt replied:
Good idea, added

Commit:
Remove bit casting of floats. Improve accuracy:

- Instead of grad_in and input, host was passing input and input
- Seems like mul_binary_tiles invalidates dest_reg[dest_index1]

Commit:
Fix bug in override_runtime_arguments that messed up accuracy

philei-tt commented:
I'm happy to inform you that bug with accuracy is finally fixed! 
Here are some logs from nano-gpt training: 

Composite gelu bw (approximation=none): (~265ms)
```
Step: 100, Loss: 2.609375
Full step time 264.138 ms, cache entries: 96
Step: 200, Loss: 2.484375
Full step time 264.74 ms, cache entries: 96
Step: 300, Loss: 2.390625
Full step time 265.854 ms, cache entries: 96
Step: 400, Loss: 2.265625
Full step time 264.017 ms, cache entries: 96
```

Fused gelu bw (approximation=none): (~

```
Step: 100, Loss: 2.59375
Full step time 238.756 ms, cache entries: 91
Step: 200, Loss: 2.484375
Full step time 238.687 ms, cache entries: 91
Step: 300, Loss: 2.375
Full step time 238.825 ms, cache entries: 91
Step: 400, Loss: 2.25
Full step time 239.163 ms, cache entries: 91
```

Commit:
Improve tanh accuracy. Fix fused gelu bw unittest

Commit:
Short comment explaining copy_value and load_immediate_value

Commit:
Improve comment explaining copy_value and load_immediate_value

philei-tt commented on tests/ttnn/unit_tests/operations/eltwise/backward/test_backward_gelu_fused.py:
```diff
@@ -0,0 +1,124 @@
+@pytest.mark.parametrize(
+    "input_shapes",
+    (
+        (torch.Size([1, 1, 32, 32])),
+        (torch.Size([1, 1, 320, 384])),
+        (torch.Size([1, 3, 320, 384])),
```
I just used same shapes that were used for composite gelu bw. Shape doesn't influence kernels in any way -- its flattened and processed in tiles.

ayerofieiev-tt replied:
@philei-tt please let’s add a test with non-tile aligned shapes, as well shapes of rank 0 and 1

philei-tt replied:
It's impossible to do rank 1 and non aligned shapes due to assert in tensor layout conversion from row-major to tile: https://github.com/tenstorrent/tt-metal/blob/2573a69beceacfd5aebc370504c86e5979acb3ff/ttnn/cpp/ttnn/tensor/tensor_impl.hpp#L92 

Shouldn't here be "auto padding" in ttnn?

ayerofieiev-tt replied:
Please align with Stas

philei-tt replied:
Hello, I have updated tests to test the following input shapes: 
```
INPUT_SHAPES = (
    (torch.Size([32])),
    (torch.Size([25, 34])),  # not aligned by tile size
    (torch.Size([32, 32])),
    (torch.Size([1, 32, 32])),
    (torch.Size([1, 1, 32, 32])),
    (torch.Size([1, 1, 320, 384])),
    (torch.Size([1, 3, 320, 384])),
    (torch.Size([1, 3, 323, 389])),  # not aligned by tile size
)
```

Commit:
Add other ranks and non aligned shapes to fused gelu bw unittest
